lib/tail-file.js

```diff
@@ -0,0 +1,279 @@
+        })
+      } else if (eof < this[kStartPos]) {
+        // Same file, but smaller/truncated
+        this[kStartPos] = 0
+        this[kInode] = stats.ino
+        fileHasChanged = true
+        this.emit('truncated', {
+          message: 'The file was truncated.  Tailing resumed from the beginning.'
+        , filename: this[kFileName]
+        , when: new Date()
+        })
+      } else if (this[kStartPos] !== eof) {
+        fileHasChanged = true
+      }
+
+      if (fileHasChanged) {
+        await this._streamFileChanges()
+      } else {
+        setImmediate(this.emit.bind(this), 'flush')
+      }
+
+      this._scheduleTimer(this[kPollFileIntervalMs])
+    } catch (err) {
+      if (err.code === 'ENOENT') {
+        if (this[kFileHandle]) {
+          // Don't lose the last lines in the file if it previously existed.
+          // Perhaps it has not been re-created yet (or won't be)
+          try {
+            await this._readRemainderFromFileHandle()
+          } catch (error) {
+            this.emit('tail_error', error)
+          }
+        }
+        this[kPollFailureCount]++
+        if (this[kPollFailureCount] >= this[kMaxPollFailures]) {
+          return this.quit(err)
+        }
+        this.emit('retry', {
+          message: 'File disappeared. Retrying.'
+        , filename: this[kFileName]
+        , attempts: this[kPollFailureCount]
+        , when: new Date()
+        })
+        this._scheduleTimer(this[kPollFailureRetryMs])
+        return
+      }
+      // Some other error like EACCES
+      // TODO: Retries for certain error codes can be put here
+      return this.quit(err)
+    }
+    return
+  }
+
+  _scheduleTimer(ms) {
+    clearTimeout(this[kPollTimer])
+    if (this[kQuitting]) return
+    this[kPollTimer] = setTimeout(this._pollFileForChanges.bind(this), ms)
+    return
+  }
+
+  async _streamFileChanges() {
+    try {
+      const stream = fs.createReadStream(this[kFileName], {
+        ...this[kOpts].readStreamOpts
+      , start: this[kStartPos]
+      })
+      await this._readChunks(stream)
+    } catch (err) {
+      // Possible file removal.  Let auto-retry handle it.
+      this[kPollFailureCount]++
+      const error = new Error('An error was encountered while tailing the file')
+      error.code = 'ETAIL'
+      error.meta = {actual: err}
+      // Emitting on 'error' would bork the parent Readstream
+      // FIXME: error can be emitted here
+      this.emit('tail_error', error)
+    }
+    return
+  }
+
+  _read() {
+    if (this[kStream]) {
+      this._readChunks(this[kStream])
+    }
+    return
+  }
+
+  async quit(err) {
+    this[kQuitting] = true
+    clearTimeout(this[kPollTimer])
+
+    if (err) {
+      this.emit('error', err)
+    } else {
+      // One last read to get lines added in high throughput
+      this._pollFileForChanges().catch(NOOP)
+      await once(this, 'flush')
+    }
+
+    // Signal the end of this Readstream
+    this.push(null)
```

Comment:
I'm not sure it matters.  What I do know is that if `error` is emitted on the Readable, it borks the entire stream, but I don't think this would fail.  Might have to verify.

Reply:
I tested this and it doesn't seem to matter.  An `error` is emitted, and so is `end`, which I think is fine.  `this.push(null)` signifies the end of the parent tail stream, so I don't see an issue with that.